### ISSUE TITLE
USB suspend not implemented in firmware

### ISSUE THREAD

**Issue by mossmann** [mossmann/hackrf#97]
Workaround here:

http://nine.pairlist.net/pipermail/hackrf-dev/2013-August/000169.html

**Comment by petterreinholdtsen** [comment]
What exactly is missing and what does it take to add it?  Or at least add enough support to reject suspend requests from the host computer?

**Comment by davidhubbard** [comment]
I've done some back-of-napkin math on suspend. It's useful to review http://www.beyondlogic.org/usbnutshell/usb2.shtml#SuspendCurrent

There are a lot of systems that don't try to auto-suspend USB devices. It's worth trying to get USB suspend working though. I don't think it's worth worrying about "sleep mode" from the Link Power Management Addendum though.

A USB device has 7ms from when the bus signals it should suspend; after that it must not draw more than [2.5mA](http://compliance.usb.org/index.asp?UpdateFile=Electrical) -- calculated from the bMaxPower of 500mA set in hackrf_usb/usb_descriptor.c.

That's not a lot of power. The 3V3 LED by itself probably uses that all up:

- 2.5mA 3V3 LED
- 6.4mA TPS62410 switcher
- 2mA (3.3V) RFFC5072 in standby mode
- 5mA (3.3V) Si5351C
- 0.03mA LPC4320 in deep sleep mode
- Maybe another 1mA in resistive losses (think: conversion losses, non-ideal components, ...).

That's 16.93mA. Happy to be corrected, though. But as far as napkin math goes, I'd say this board will draw more than 2.5mA, best case.

I studied the schematic and see that it would be easy to assert VBUSCTRL = 5V, and connect a power supply to P20. Actual current measurements might be helpful.

**Comment by petterreinholdtsen** [comment]
[David Hubbard]
> I've done some back-of-napkin math on suspend. It's useful to review
> http://www.beyondlogic.org/usbnutshell/usb2.shtml#SuspendCurrent
>
> There are a lot of systems that don't try to auto-suspend USB
> devices. It's worth trying to get USB suspend working though. I don't
> think it's worth worrying about "sleep mode" from the Link Power
> Management Addendum though.

If suspending is too hard, is it possible to respond to a suspend
request and reject it?  Would that allow the computer to know that
suspend do not work and hopefully get the HackRF One working on
Thinkpads?

-- 
Happy hacking
Petter Reinholdtsen

**Comment by davidhubbard** [comment]
> If suspending is too hard, is it possible to respond to a suspend
request and reject it?  Would that allow the computer to know that
suspend do not work and hopefully get the HackRF One working on
Thinkpads?

Sorry, a device cannot block a host that wants to suspend it.

USB suspend is defined at the USB physical layer. The device must enter suspend when the host stops sending anything over the bus. Yep, that's how suspend is signalled.

Since USB is a master/slave protocol, there is nothing a device can do when the host stops talking to reject the suspend request. In fact, doing nothing is actually an ok response, as the host should then detect that the device is still drawing current and cut power to the device.

**Comment by dominicgs** [comment]
@petterreinholdtsen Does the workaround not work for ThinkPads?  I thought one of the two work around methods worked on all systems so far.

**Comment by petterreinholdtsen** [comment]
[Dominic Spill]
> @petterreinholdtsen Does the workaround not work for ThinkPads?  I
> thought one of the two work around methods worked on all systems so
> far.

Not reliably on my laptop, at least.  I need to try several times before
I am able to get it to work.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by davidhubbard** [comment]
Really the HackRF shouldn't have any problem with suspend. I'd like to at least try some tests to find out what it does when the USB port goes into suspend. My hunch is that the problem is: (1) the HackRF continues to operate as if it had full power, (2) the USB port on the host loses power in an undefined way, (3) voltages on the HackRF board "brown out", (4) something goes into a tri-stated or unresponsive state even when power is restored.

In a few weeks I'll run some tests and see if I can reproduce the problem. It should then be possible to detect the suspend even though the HackRF can't shut off all the uses of power. The HackRF should be able to "stop operations" such that it can survive a brown out. Or, if it's not a brown out that's happening, maybe I can post more details.

**Comment by davidhubbard** [comment]
I ran some tests but cannot reproduce the problem. Here's what I get:

```
$ sudo -s
# MY_DEV=/sys/bus/usb/devices/usb2/2-3
# cat $MY_DEV/power/runtime_status
active
# echo auto > $MY_DEV/power/control
# cat $MY_DEV/power/runtime_status
suspended
# echo on > $MY_DEV/power/control
# cat $MY_DEV/power/runtime_status
active
# exit
$ hackrf_info
```
And hackrf_info works right off. Looking at the output of dmesg, I see:
```
usb 2-3: reset high-speed USB device number 27 using xhci_hcd
usb 2-3: USB disconnect, device number 27
usb 2-3: new high-speed USB device number 29 using xhci_hcd
```
Which is a clue. My xhci controller is disconnecting the device entirely, which may be why I don't see the problem when trying to get the device out of suspend.

I will see if I can find a different USB port to work with.

**Comment by straithe** [comment]
This issue was resolved in release [2021.03.1](https://github.com/mossmann/hackrf/releases/tag/v2021.03.1).

**Comment by petterreinholdtsen** [comment]
[Straithe]
> This issue was resolved in release
> [2021.03.1](https://github.com/mossmann/hackrf/releases/tag/v2021.03.1).

Thank you very much for letting us know.

-- 
Happy hacking
Petter Reinholdtsen
